### ISSUE TITLE
Coverage for native has recently broken

### ISSUE THREAD

**Issue by chrisseaton** [oracle/truffleruby#1994]
```
% tool/jt.rb --use native ruby --coverage -e 14
Using Native TruffleRuby with Graal: mxbuild/truffleruby-native
$ /Users/chrisseaton/src/github.com/Shopify/truffleruby-shopify/mxbuild/truffleruby-native/jre/languages/ruby/bin/truffleruby --coverage -e 14
truffleruby: an exception escaped out of the interpreter - this is an implementation bug
org.graalvm.polyglot.PolyglotException: java.lang.NullPointerException
	at com.oracle.truffle.polyglot.EngineAccessor$EngineImpl.getCurrentContext(EngineAccessor.java:347)
	at com.oracle.truffle.api.TruffleLanguage.getCurrentContext(TruffleLanguage.java:1485)
	at org.truffleruby.RubyLanguage.getCurrentContext(RubyLanguage.java:149)
	at org.truffleruby.language.RubyContextSourceNode.getContext(RubyContextSourceNode.java:69)
	at org.truffleruby.language.LazyRubyNode.getOrCreateMasterResolution(LazyRubyNode.java:93)
	at org.truffleruby.language.LazyRubyNode.lambda$resolve$0(LazyRubyNode.java:76)
	at com.oracle.truffle.api.nodes.Node.atomic(Node.java:577)
	at org.truffleruby.language.LazyRubyNode.resolve(LazyRubyNode.java:71)
	at org.truffleruby.language.LazyRubyNode.materializeInstrumentableNodes(LazyRubyNode.java:57)
	at com.oracle.truffle.api.instrumentation.InstrumentationHandler$AbstractNodeVisitor.materializeSyntaxNodes(InstrumentationHandler.java:1668)
	at com.oracle.truffle.api.instrumentation.InstrumentationHandler$AbstractNodeVisitor.materialize(InstrumentationHandler.java:1573)
	at com.oracle.truffle.api.instrumentation.InstrumentationHandler$AbstractNodeVisitor.visit(InstrumentationHandler.java:1534)
	at com.oracle.truffle.api.nodes.NodeUtil.forEachChild(NodeUtil.java:468)
	at com.oracle.truffle.api.instrumentation.InstrumentationHandler$AbstractNodeVisitor.visit(InstrumentationHandler.java:1561)
	at com.oracle.truffle.api.instrumentation.InstrumentationHandler.visitRoot(InstrumentationHandler.java:1204)
	at com.oracle.truffle.api.instrumentation.InstrumentationHandler.visitRoots(InstrumentationHandler.java:747)
	at com.oracle.truffle.api.instrumentation.InstrumentationHandler.addSourceSectionBinding(InstrumentationHandler.java:486)
	at com.oracle.truffle.api.instrumentation.InstrumentationHandler.attachSourceSectionListener(InstrumentationHandler.java:1043)
	at com.oracle.truffle.api.instrumentation.InstrumentationHandler.access$2700(InstrumentationHandler.java:93)
	at com.oracle.truffle.api.instrumentation.InstrumentationHandler$AbstractInstrumenter.attachLoadSourceSectionListener(InstrumentationHandler.java:2170)
	at com.oracle.truffle.tools.coverage.CoverageTracker.instrumentLoadedStatements(CoverageTracker.java:279)
	at com.oracle.truffle.tools.coverage.CoverageTracker.instrument(CoverageTracker.java:256)
	at com.oracle.truffle.tools.coverage.CoverageTracker.start(CoverageTracker.java:143)
	at com.oracle.truffle.tools.coverage.impl.CoverageInstrument.onCreate(CoverageInstrument.java:167)
	at com.oracle.truffle.api.instrumentation.TruffleInstrument$Env.onCreate(TruffleInstrument.java:408)
	at com.oracle.truffle.api.instrumentation.InstrumentationHandler$InstrumentClientInstrumenter.create(InstrumentationHandler.java:1856)
	at com.oracle.truffle.api.instrumentation.InstrumentationHandler.createInstrument(InstrumentationHandler.java:396)
	at com.oracle.truffle.api.instrumentation.InstrumentAccessor$InstrumentImpl.createInstrument(InstrumentAccessor.java:103)
	at com.oracle.truffle.polyglot.PolyglotInstrument.ensureCreated(PolyglotInstrument.java:128)
	at com.oracle.truffle.polyglot.PolyglotEngineImpl.ensureInstrumentsCreated(PolyglotEngineImpl.java:425)
	at com.oracle.truffle.polyglot.PolyglotEngineImpl.loadPreinitializedContext(PolyglotEngineImpl.java:1498)
	at com.oracle.truffle.polyglot.PolyglotEngineImpl.createContext(PolyglotEngineImpl.java:1442)
	at org.graalvm.polyglot.Context$Builder.build(Context.java:1490)
	at org.truffleruby.launcher.RubyLauncher.createContext(RubyLauncher.java:265)
	at org.truffleruby.launcher.RubyLauncher.runRubyMain(RubyLauncher.java:204)
	at org.truffleruby.launcher.RubyLauncher.launch(RubyLauncher.java:123)
	at org.graalvm.launcher.AbstractLanguageLauncher.launch(AbstractLanguageLauncher.java:121)
	at org.graalvm.launcher.AbstractLanguageLauncher.launch(AbstractLanguageLauncher.java:70)
	at org.truffleruby.launcher.RubyLauncher.main(RubyLauncher.java:38)
Original Internal Error: 
java.lang.NullPointerException
	at com.oracle.truffle.polyglot.EngineAccessor$EngineImpl.getCurrentContext(EngineAccessor.java:347)
	at com.oracle.truffle.api.TruffleLanguage.getCurrentContext(TruffleLanguage.java:1485)
	at org.truffleruby.RubyLanguage.getCurrentContext(RubyLanguage.java:149)
	at org.truffleruby.language.RubyContextSourceNode.getContext(RubyContextSourceNode.java:69)
	at org.truffleruby.language.LazyRubyNode.getOrCreateMasterResolution(LazyRubyNode.java:93)
	at org.truffleruby.language.LazyRubyNode.lambda$resolve$0(LazyRubyNode.java:76)
	at com.oracle.truffle.api.nodes.Node.atomic(Node.java:577)
	at org.truffleruby.language.LazyRubyNode.resolve(LazyRubyNode.java:71)
	at org.truffleruby.language.LazyRubyNode.materializeInstrumentableNodes(LazyRubyNode.java:57)
	at com.oracle.truffle.api.instrumentation.InstrumentationHandler$AbstractNodeVisitor.materializeSyntaxNodes(InstrumentationHandler.java:1668)
	at com.oracle.truffle.api.instrumentation.InstrumentationHandler$AbstractNodeVisitor.materialize(InstrumentationHandler.java:1573)
	at com.oracle.truffle.api.instrumentation.InstrumentationHandler$AbstractNodeVisitor.visit(InstrumentationHandler.java:1534)
	at com.oracle.truffle.api.nodes.NodeUtil.forEachChild(NodeUtil.java:468)
	at com.oracle.truffle.api.instrumentation.InstrumentationHandler$AbstractNodeVisitor.visit(InstrumentationHandler.java:1561)
	at com.oracle.truffle.api.instrumentation.InstrumentationHandler.visitRoot(InstrumentationHandler.java:1204)
	at com.oracle.truffle.api.instrumentation.InstrumentationHandler.visitRoots(InstrumentationHandler.java:747)
	at com.oracle.truffle.api.instrumentation.InstrumentationHandler.addSourceSectionBinding(InstrumentationHandler.java:486)
	at com.oracle.truffle.api.instrumentation.InstrumentationHandler.attachSourceSectionListener(InstrumentationHandler.java:1043)
	at com.oracle.truffle.api.instrumentation.InstrumentationHandler.access$2700(InstrumentationHandler.java:93)
	at com.oracle.truffle.api.instrumentation.InstrumentationHandler$AbstractInstrumenter.attachLoadSourceSectionListener(InstrumentationHandler.java:2170)
	at com.oracle.truffle.tools.coverage.CoverageTracker.instrumentLoadedStatements(CoverageTracker.java:279)
	at com.oracle.truffle.tools.coverage.CoverageTracker.instrument(CoverageTracker.java:256)
	at com.oracle.truffle.tools.coverage.CoverageTracker.start(CoverageTracker.java:143)
	at com.oracle.truffle.tools.coverage.impl.CoverageInstrument.onCreate(CoverageInstrument.java:167)
	at com.oracle.truffle.api.instrumentation.TruffleInstrument$Env.onCreate(TruffleInstrument.java:408)
	at com.oracle.truffle.api.instrumentation.InstrumentationHandler$InstrumentClientInstrumenter.create(InstrumentationHandler.java:1856)
	at com.oracle.truffle.api.instrumentation.InstrumentationHandler.createInstrument(InstrumentationHandler.java:396)
	at com.oracle.truffle.api.instrumentation.InstrumentAccessor$InstrumentImpl.createInstrument(InstrumentAccessor.java:103)
	at com.oracle.truffle.polyglot.PolyglotInstrument.ensureCreated(PolyglotInstrument.java:128)
	at com.oracle.truffle.polyglot.PolyglotEngineImpl.ensureInstrumentsCreated(PolyglotEngineImpl.java:425)
	at com.oracle.truffle.polyglot.PolyglotEngineImpl.loadPreinitializedContext(PolyglotEngineImpl.java:1498)
	at com.oracle.truffle.polyglot.PolyglotEngineImpl.createContext(PolyglotEngineImpl.java:1442)
	at org.graalvm.polyglot.Context$Builder.build(Context.java:1490)
	at org.truffleruby.launcher.RubyLauncher.createContext(RubyLauncher.java:265)
	at org.truffleruby.launcher.RubyLauncher.runRubyMain(RubyLauncher.java:204)
	at org.truffleruby.launcher.RubyLauncher.launch(RubyLauncher.java:123)
	at org.graalvm.launcher.AbstractLanguageLauncher.launch(AbstractLanguageLauncher.java:121)
	at org.graalvm.launcher.AbstractLanguageLauncher.launch(AbstractLanguageLauncher.java:70)
	at org.truffleruby.launcher.RubyLauncher.main(RubyLauncher.java:38)
Caused by: Attached Guest Language Frames (0)
```

**Comment by eregon** [comment]
Thanks for the report.
We need that `LazyRubyNode#materializeInstrumentableNodes` to get line breakpoints to work in the debugger when the breakpoint is set before the first execution of that method.

`Node#materializeInstrumentableNodes()` is documented as `The method may be invoked without a language context currently being active.`

And like a good part of Ruby parsing/translation that needs the RubyContext (although we could probably pass it more directly here).

During pre-initialization, the instruments are created without the context being entered. That's maybe something that could change (patchContext() is run with the context entered just before, but not creating the instruments currently). cc @tzezula @chumer

**Comment by eregon** [comment]
Fixing the obvious usage of `getContext()` in LazyRubyNode doesn't solve it:
```
java.lang.NullPointerException
	at com.oracle.truffle.polyglot.EngineAccessor$EngineImpl.getCurrentContext(EngineAccessor.java:347)
	at com.oracle.truffle.api.TruffleLanguage.getCurrentContext(TruffleLanguage.java:1485)
	at org.truffleruby.RubyLanguage.getCurrentContext(RubyLanguage.java:149)
	at org.truffleruby.language.RubyContextSourceNode.getContext(RubyContextSourceNode.java:69)
	at org.truffleruby.core.inlined.InlinedOperationNode.<init>(InlinedOperationNode.java:37)
	at org.truffleruby.core.inlined.UnaryInlinedOperationNode.<init>(UnaryInlinedOperationNode.java:22)
	at org.truffleruby.core.inlined.InlinedNotNode.<init>(InlinedNotNode.java:26)
	at org.truffleruby.core.inlined.InlinedNotNodeGen.<init>(InlinedNotNodeGen.java:28)
	at org.truffleruby.core.inlined.InlinedNotNodeGen.create(InlinedNotNodeGen.java:127)
	at org.truffleruby.core.inlined.CoreMethods.createCallNode(CoreMethods.java:161)
	at org.truffleruby.parser.BodyTranslator.translateCallNode(BodyTranslator.java:599)
	at org.truffleruby.parser.BodyTranslator.visitCallNode(BodyTranslator.java:516)
	at org.truffleruby.parser.BodyTranslator.visitCallNode(BodyTranslator.java:269)
```
And passing the Context all the way down there is not nice.

**Comment by eregon** [comment]
Also coverage is a use-case that shouldn't require early materialization like this, but rather only instrument on first execution with `notifyInserted()`. Maybe materializeInstrumentableNodes() should only be called by tools really needing it that eagerly and not by default when instrumenting?
cc @entlicher 

Not clear to me what's a good fix here.

**Comment by chrisseaton** [comment]
know what caused the regression though? It wasn't here last week.

**Comment by eregon** [comment]
It's 867e3a8c7fcff3bfff4ecaa2f8bc841d48f2a685 part of a8b9b9cc6da089b9127647b00255bdd73473c11c, which is needed to fix a RubyDebugTest.

**Comment by eregon** [comment]
I ended up passing the RubyContext to all nodes needing it in the constructor, and adding a spec that `--coverage` works. It was quite a few nodes to adapt but it's done now.

**Comment by eregon** [comment]
Fixed in 91e7ba048ab2bb9f71bddc028da281f6b651d0ca
